third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -1005,22 +1015,32 @@ impl<'env, 'state> LifetimeAnalysisStep<'env, 'state> {
         }
     }
 
-    /// Check whether a local can be written. This is only allowed if no borrowed references exist.
+    /// Check whether a local can be written. This is only allowed if it is not borrowed by any
+    /// temps that are alive after the statement
     fn check_write_local(&self, local: TempIndex) {
         if self.is_ref(local) {
             // Always valid
             return;
         }
         if let Some(label) = self.state.label_for_temp_with_children(local) {
-            // The destination is currently borrowed and cannot be assigned
-            self.error_with_hints(
-                self.cur_loc(),
-                format!("cannot assign to borrowed {}", self.display(local)),
-                "attempted to assign here",
-                self.borrow_info(label, |_| true)
-                    .into_iter()
-                    .chain(self.usage_info(label, |t| t != &local)),
-            )
+            let children = self.state.transitive_children(label);
+            // cannot be assigned if the destination is currently borrowed by temps that are alive
+            // after the statement
+            if self
+                .state
+                .leaves_for_labels(&children)
+                .iter()
+                .any(|temp| self.alive.after.contains_key(temp) && *temp != local)
+            {
```

Comment:
This looks sound, however, I wonder why this is needed. There seems to be a more deeper bug which triggers this treatment.

In general, the invariant holds that if something is in the `temp_to_label` map, its also directly or indirectly alive. 

So why is then `local` in the map, in the case none of the children are?

Reply:
The original looks like this:
```
...
  1: $t2 := borrow_local($t0)
  2: ($t0, $t1) := M::foo($t2)
...  
```
$2 is borrowed from $0 but $0 is reassigned in this function call so the current check will raise an error. In this revised check, we first try to find if there are any temps borrowing local after this program point. If not, we don't generate an error for it.

Reply:
The issue is that t2 is going out of scope in the same step where it is used, so at the moment we assign t0, it is actually not used any more. Your fix is not the right way to deal with this, rather we want to fix this more general. One thing we could do is pull the cleanup we do after and before each instruction in `execute` into its own function and call that function before we make the check for the assigned destination. We'd want to do the same for other places were we assign variables.

Reply:
Added the logic to release references.